File: engine/internal/workload/promote.go
package workload

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"

	aferrors "github.com/antifailure/antifailure/engine/internal/errors"
	"github.com/antifailure/antifailure/engine/internal/explore"
	"github.com/antifailure/antifailure/engine/pkg/schema"
)

// Promotion turns a discovery into a check that runs on every pull request.
//
// THE THING THIS FILE REFUSES TO PRETEND.
//
// A compiled workflow does not replay the exploration. The explorer wandered,
// found a route, and wrote down where it went; the workflow that comes out of
// that is handed to a planner which works out its own route from the same
// starting point towards the same goal. Two runs of the promoted workflow can
// therefore take different paths, and the day the application changes, the
// promoted workflow will take a third one and still pass.
//
// That is not a defect to fix here. It is what makes the workflow survive a
// redesign, which is the whole reason a declared workflow is worth more than a
// recorded one. But a promotion that did not SAY it would be a lie by omission,
// because a person reading "promoted from the exploration that found this" will
// reasonably assume the two do the same thing.
//
// So every promotion carries three things nothing else in this product carries:
// the list of what compilation could not take with it, spelled out one sentence
// each; a digest of the journey the exploration actually walked, so a later
// exploration from the same seed can be compared against it and drift can be
// seen rather than guessed at; and the exact af explore command that produced
// the discovery, so somebody can go and look.

// PromotionSchema names the wire format.
const PromotionSchema = "antifailure.workload.promotion/v1"

// Promotion is a discovery, compiled, with everything the compilation dropped.
type Promotion struct {
	Schema string `json:"schema"`
	// Kind is always browser_workflow. An exploration compiles into a declared
	// workflow and into nothing else: it drives a browser, so it cannot become
	// an HTTP scenario, and it has no request rate, so it cannot become a mix.
	Kind Kind          `json:"kind"`
	Body PromotionBody `json:"body"`
	// BodyDigest is sha256 of the body in the canonical form the control plane
	// uses: keys sorted at every depth, absent fields omitted. It is what
	// answers "is this the same definition" for a promotion that would
	// otherwise write a duplicate version.
	BodyDigest string          `json:"body_digest"`
	Source     PromotionSource `json:"source"`
	// Dropped is what the compilation could not carry over, one sentence each.
	// Never empty, because there is always something.
	Dropped []string `json:"dropped"`
	// Notes are the compiler's own remarks about this exploration, which are
	// about what it saw rather than about what compilation loses.
	Notes []string `json:"notes"`
}

// PromotionBody is the version body the control plane stores, in exactly the
// shape its browser_workflow schema declares.
type PromotionBody struct {
	// Select is the workflow name, which is the exploration's name. It finds
	// nothing until the manifest block below is pasted into the repository,
	// which is why the block is part of the body rather than a note beside it.
	Select []string `json:"select"`
	// ManifestBlock is the YAML a person adds to antifailure.yaml.
	ManifestBlock string `json:"manifestBlock"`
	// Dropped is the same list as the promotion's, carried into the body
	// because it is part of what the version IS. A version that lost it would
	// read as a faithful recording a release later.
	Dropped []string `json:"dropped"`
}

// PromotionSource is where the discovery came from and how to go back to it.
type PromotionSource struct {
	Exploration string `json:"exploration"`
	Goal        string `json:"goal"`
	Seed        string `json:"seed"`
	Reached     bool   `json:"reached"`
	// JourneySteps and JourneyDigest describe the path the exploration walked.
	//
	// The digest is the drift anchor. Re-running the same goal from the same
	// seed against an unchanged application walks the same path and produces
	// the same digest; a different digest means the application moved under
	// the promoted workflow, which is the moment somebody should look at it
	// rather than the moment a green check quietly stops meaning anything.
	JourneySteps  int    `json:"journey_steps"`
	JourneyDigest string `json:"journey_digest"`
	// Findings are what the exploration noticed on the way. They are recorded
	// and deliberately not turned into expectations: "pressing Upgrade plan
	// changes nothing" is a defect to fix, and a workflow asserting it would
	// go green the day somebody broke it differently.
	Findings []string `json:"findings,omitempty"`
	// Reproduce is the af explore command that found this.
	Reproduce Reproduce `json:"reproduce"`
}

// droppedByCompilation is what a promotion always loses, stated once.
//
// A list rather than a paragraph because it goes into a version body and into
// a console, and because each line is a separate thing somebody might
// reasonably have assumed survived.
func droppedByCompilation(e explore.Exploration) []string {
	out := []string{
		"the workflow is planned again from the start path rather than replayed, " +
			"so it can take a different route to the same goal and still pass",
		"the values the exploration typed into forms are not carried over; the " +
			"workflow's persona supplies its own",
		"the exploration's seed does not steer the workflow, because a declared " +
			"workflow makes no random choices to seed",
		"the pages the exploration visited on the way are not asserted, only the " +
			"goal it was looking for",
	}
	if n := len(e.Findings); n > 0 {
		noun := "friction findings are"
		if n == 1 {
			noun = "friction finding is"
		}
		out = append(out, fmt.Sprintf(
			"%d %s recorded and not asserted, because a defect to fix is not an "+
				"outcome to require", n, noun))
	}
	if e.Evidence.Trace != "" || e.Evidence.Video != "" || e.Evidence.Screenshot != "" {
		out = append(out, "the trace, video and screenshot belong to the exploration's own "+
			"run and are not carried into the workflow")
	}
	return out
}

// Promote compiles one exploration into a versioned workflow definition.
//
// It refuses an exploration that did not reach its goal, and that is stricter
// than af explore --emit-workflow, deliberately. Printing a block for somebody
// to read and edit is a different act from writing a version a hosted run will
// execute unattended: the expectation a compiled workflow asserts is the goal
// sentence, and a wander that never got there has no evidence the goal is
// reachable at all. Promoting one would create a check that has never passed
// and cannot say why.
func Promote(e explore.Exploration, persona string, seed string) (*Promotion, error) {
	switch {
	case e.Outcome.Verdict == VerdictBlocked:
		return nil, aferrors.Coded(aferrors.AFWLD010, "exploration", e.Name,
			"detail", "the exploration was blocked, so it has no journey to compile")
	case !e.Reached:
		return nil, aferrors.Coded(aferrors.AFWLD010, "exploration", e.Name,
			"detail", "the exploration did not reach its goal, so the workflow it would "+
				"compile into would assert something nothing has ever shown to be reachable")
	case strings.TrimSpace(e.Name) == "":
		return nil, aferrors.Coded(aferrors.AFWLD010, "exploration", "(unnamed)",
			"detail", "an exploration with no name compiles into a workflow nothing can select")
	}

	workflow, notes := explore.Compile(e, persona)
	block, err := manifestBlock(workflow)
	if err != nil {
		return nil, aferrors.Coded(aferrors.AFWLD010, "exploration", e.Name,
			"detail", "the compiled workflow could not be written as YAML: "+err.Error())
	}

	dropped := droppedByCompilation(e)
	body := PromotionBody{
		Select:        []string{workflow.Name},
		ManifestBlock: block,
		Dropped:       dropped,
	}

	replay := Plan{Kind: Exploration, Select: []string{e.Name}, SeedText: orSeed(seed, e.Seed)}
	p := &Promotion{
		Schema:     PromotionSchema,
		Kind:       BrowserWorkflow,
		Body:       body,
		BodyDigest: BodyDigest(body),
		Source: PromotionSource{
			Exploration:   e.Name,
			Goal:          e.Goal,
			Seed:          e.Seed,
			Reached:       e.Reached,
			JourneySteps:  len(e.Journey),
			JourneyDigest: JourneyDigest(e),
			Findings:      findingLines(e),
			Reproduce: Reproduce{
				Argv:    replay.Argv(),
				Command: replay.Command(),
				Note: "run this from the repository root to walk the same goal from the " +
					"same seed and compare the journey digest",
			},
		},
		Dropped: dropped,
		Notes:   notes,
	}
	return p, nil
}

func orSeed(preferred, fallback string) string {
	if strings.TrimSpace(preferred) != "" {
		return preferred
	}
	return fallback
}

func findingLines(e explore.Exploration) []string {
	out := make([]string, 0, len(e.Findings))
	for _, f := range e.Findings {
		out = append(out, fmt.Sprintf("%s at %s: %s", f.Kind, f.URL, f.Detail))
	}
	return out
}

// JourneyDigest is a stable fingerprint of the path an exploration walked.
//
// The moves rather than the pages, because a page whose address carries a
// session identifier differs between two runs of an unchanged application and
// the moves do not. Values typed into fields are excluded for the same reason:
// a persona's generated address changes per run and would make every journey
// look different from every other.
func JourneyDigest(e explore.Exploration) string {
	h := sha256.New()
	for _, m := range e.Journey {
		_, _ = fmt.Fprintf(h, "%s\x00%s\x00%s\x00%s\n", m.Kind, m.URL, m.Field, m.Control)
	}
	return hex.EncodeToString(h.Sum(nil))
}

// Drift compares a promotion against a fresh exploration of the same goal.
//
// The question it answers is the one a promotion cannot answer on its own: is
// the check still checking what it was promoted for. A promoted workflow is
// planned again on every run, so it goes on passing while the route it was
// found on disappears, and nothing about the workflow's own result would say
// so.
type Drift struct {
	Exploration string `json:"exploration"`
	// SameSeed says whether the two were walked from the same seed. A drift
	// report over two different seeds says nothing, so this is reported rather
	// than assumed.
	SameSeed bool `json:"same_seed"`
	// Moved is true when the journey digests differ.
	Moved bool `json:"moved"`
	// StillReaches says whether the fresh exploration got to the goal at all,
	// which is the more serious of the two answers.
	StillReaches bool   `json:"still_reaches"`
	WasSteps     int    `json:"was_steps"`
	NowSteps     int    `json:"now_steps"`
	WasDigest    string `json:"was_digest"`
	NowDigest    string `json:"now_digest"`
	Detail       string `json:"detail"`
}

// CompareJourney reports whether an exploration has moved since it was
// promoted.
func CompareJourney(p Promotion, now explore.Exploration) Drift {
	d := Drift{
		Exploration:  p.Source.Exploration,
		SameSeed:     p.Source.Seed == now.Seed,
		StillReaches: now.Reached,
		WasSteps:     p.Source.JourneySteps,
		NowSteps:     len(now.Journey),
		WasDigest:    p.Source.JourneyDigest,
		NowDigest:    JourneyDigest(now),
	}
	d.Moved = d.WasDigest != d.NowDigest
	switch {
	case !d.SameSeed:
		d.Detail = "the two explorations were walked from different seeds, so a difference " +
			"between them says nothing about the application"
	case !now.Reached:
		d.Detail = "the goal is no longer reachable from this seed, and the promoted workflow " +
			"is planned again on every run so it can still be passing"
	case d.Moved:
		d.Detail = "the route to the goal has changed since this workflow was promoted"
	default:
		d.Detail = "the route to the goal is unchanged"
	}
	return d
}

// manifestBlock renders the workflow as the YAML somebody pastes.
func manifestBlock(w schema.Workflow) (string, error) {
	body, err := yaml.Marshal(struct {
		Workflows []schema.Workflow `yaml:"workflows"`
	}{Workflows: []schema.Workflow{w}})
	if err != nil {
		return "", err
	}
	return string(body), nil
}

// BodyDigest is sha256 of a version body in the control plane's canonical
// form.
//
// Canonical means keys sorted at every depth and absent fields omitted, which
// is what the control plane's own digestOf does. Two implementations of the
// same digest is a place to drift, so the shape is stated here rather than
// inferred: an object is {"key":value} with keys sorted bytewise, an array is
// [a,b] in order, and everything else is its JSON literal. A round trip test
// against a fixture keeps the two honest.
func BodyDigest(b PromotionBody) string {
	sum := sha256.Sum256([]byte(canonicalJSON(bodyMap(b))))
	return hex.EncodeToString(sum[:])
}

func bodyMap(b PromotionBody) map[string]any {
	m := map[string]any{
		"select":        toAnySlice(b.Select),
		"manifestBlock": b.ManifestBlock,
	}
	// Omitted rather than written as an empty array, because the control plane
	// treats an absent optional field and an empty one as the same definition
	// and its canonical form drops undefined.
	if len(b.Dropped) > 0 {
		m["dropped"] = toAnySlice(b.Dropped)
	}
	return m
}

func toAnySlice(in []string) []any {
	out := make([]any, 0, len(in))
	for _, s := range in {
		out = append(out, s)
	}
	return out
}

func canonicalJSON(v any) string {
	switch t := v.(type) {
	case map[string]any:
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			key, _ := json.Marshal(k)
			parts = append(parts, string(key)+":"+canonicalJSON(t[k]))
		}
		return "{" + strings.Join(parts, ",") + "}"
	case []any:
		parts := make([]string, 0, len(t))
		for _, e := range t {
			parts = append(parts, canonicalJSON(e))
		}
		return "[" + strings.Join(parts, ",") + "]"
	default:
		body, err := json.Marshal(v)
		if err != nil {
			return "null"
		}
		return string(body)
	}
}

File: web/apps/api/src/workloads/promote.ts
// Turning an exploration into a workflow that runs every time, and being
// honest about the half that does not survive.
//
// WHAT PROMOTION ACTUALLY IS HERE, BECAUSE IT IS EASY TO OVERSELL.
//
// An exploration is a goal and a seed. The runner reads each page, chooses
// where to go, and records the moves it made and the places the application
// cost somebody effort. `engine/internal/explore.Compile` already turns that
// into the manifest block that replays it, and this is the same compilation
// with the same two admissions, written on the side that can version it.
//
// It does NOT produce a load scenario. The audit that started this workstream
// says so plainly and it is worth repeating where somebody might assume
// otherwise: an exploration emits a MANIFEST WORKFLOW. There is no path from a
// wander through a browser to a weighted traffic mix, because the two measure
// different things and nothing in the exploration record carries a rate.
//
// WHAT SURVIVES:
//
//   the name, so the workflow can be selected by it
//   the start path, taken from the first place the browser actually went
//   the journey, as a description in accessible names rather than selectors
//   a step budget with room above what the exploration spent
//   the goal, as the expectation
//
// WHAT DOES NOT, AND WHY EACH IS DROPPED RATHER THAN GUESSED AT:
//
//   The expectation is the goal sentence and not a passing page's words. An
//   exploration knows what it was looking for and does not know what a passing
//   run should say. A workflow whose expectation cannot be read comes back
//   `unverified`, which is the honest answer and not a pass.
//
//   A friction finding is not an expectation. "Pressing Upgrade plan changes
//   nothing" is a defect to fix, not an outcome to assert: a workflow that
//   asserted it would go green the day somebody broke it differently and red
//   the day somebody fixed it.
//
//   The unexplored parts of the application are named and not covered. The
//   compiled workflow walks one route.
//
//   An exploration that never reached its goal still compiles, and the
//   workflow then asserts something nobody has seen happen. That is said out
//   loud rather than silently produced.
//
// AND THE ONE THIS SIDE ADDS, WHICH THE ENGINE'S COMPILER DOES NOT HAVE TO
// SAY: the block has to be pasted into the repository's antifailure.yaml before
// anything can run it. `af test --only <name>` selects out of the manifest, and
// the control plane cannot put a file in somebody's repository. So a promoted
// version is a definition plus an instruction, and the route returns both.

/** What a promotion produced. */
export interface Compiled {
  slug: string
  name: string
  description: string
  /** The browser_workflow body: what to select once the block is in place. */
  body: { select: string[]; manifestBlock: string; dropped: string[] }
  /** One sentence per thing the compilation deliberately did not carry. */
  dropped: string[]
  /** The block to paste into antifailure.yaml. */
  manifestBlock: string
}

export class ExplorationRefused extends Error {}

/** How many journey steps a description enumerates before it says "and n more".
 *  The same twenty the engine's compiler uses, and for the same reason: a forty
 *  step exploration would spend the whole description on a list nobody reads
 *  past the tenth line. */
const MAX_JOURNEY_LINES = 20

interface Move {
  kind: string
  url?: string
  field?: string
  control?: string
}

interface Finding {
  kind: string
  url?: string
  control?: string
}

/**
 * Compiles an `af explore --json` exploration.
 *
 * Tolerant on every list and strict on the two fields that decide what the
 * workflow IS. A journey entry that cannot be read is skipped rather than
 * throwing, because one malformed move must not discard a whole walk; a missing
 * name or goal is refused, because a workflow with neither is not a workflow.
 */
export function compileExploration(raw: unknown, persona: string | null): Compiled {
  const e = object(raw)
  const name = text(e.name, 200)
  const goal = text(e.goal, 2000)
  if (name === null || goal === null) {
    throw new ExplorationRefused(
      'an exploration has to carry a name and a goal; this document carries neither, so there ' +
        'is nothing to compile into a workflow',
    )
  }
  const slug = slugify(name)
  if (slug === null) {
    throw new ExplorationRefused(
      `${name} has no letters or digits in it, so there is no name a manifest could carry`,
    )
  }

  const seed = text(e.seed, 200)
  const reached = e.reached === true
  const journey = moves(e.journey)
  const findings = findingList(e.findings)
  const missing = strings(e.missing, 100, 300)

  const dropped: string[] = [
    'The expectation is the goal, because an exploration knows what it was looking for and not ' +
      'what a passing page should say. Check its words appear on the page the run ends on, or ' +
      'rewrite it: a workflow whose expectation cannot be read comes back unverified rather ' +
      'than as a pass.',
  ]
  if (!reached) {
    dropped.push(
      'This exploration never reached the goal, so the workflow asserts something nobody has ' +
        'seen happen. Expect it to be unverified until the path exists.',
    )
  }
  for (const f of findings) {
    if (f.kind === 'goal_unreached') continue
    const where = f.control ? `${JSON.stringify(f.control)} on ${f.url ?? 'a page'}` : (f.url ?? 'a page')
    dropped.push(
      `The exploration found ${friction(f.kind)} at ${where}, which this workflow will not ` +
        `assert. A friction finding is something to fix, not an outcome to expect.`,
    )
  }
  if (missing.length > 0) {
    dropped.push(
      `${missing.length} ${missing.length === 1 ? 'part' : 'parts'} of the application ` +
        `${missing.length === 1 ? 'was' : 'were'} left unexplored. The compiled workflow covers ` +
        `only the route that was walked.`,
    )
  }
  dropped.push(
    'Nothing runs this until the block below is in the repository. `af test --only` selects out ' +
      'of antifailure.yaml, and a control plane cannot put a file in somebody else’s ' +
      'repository.',
  )

  const description = describe(goal, seed, reached, journey)
  const manifestBlock = block({
    name: slug,
    description,
    persona,
    startPath: startPath(journey),
    expect: sentence(goal),
    steps: journey.length + 10,
  })

  return {
    slug,
    name,
    description,
    body: { select: [slug], manifestBlock, dropped },
    dropped,
    manifestBlock,
  }
}

/** The manifest block, rendered by hand rather than through a YAML library.
 *
 *  Deliberately: the control plane has no YAML dependency, this block is four
 *  keys and a list, and the one thing a library would buy, escaping, is bought
 *  here by quoting every scalar with JSON.stringify, which produces a string a
 *  YAML parser reads identically. A library would also reflow and re-quote what
 *  it was given, and this text is pasted into somebody's file. */
function block(w: {
  name: string
  description: string
  persona: string | null
  startPath: string
  expect: string
  steps: number
}): string {
  const lines = [
    'workflows:',
    `  - name: ${JSON.stringify(w.name)}`,
    `    description: ${JSON.stringify(w.description)}`,
  ]
  if (w.persona) lines.push(`    persona: ${JSON.stringify(w.persona)}`)
  lines.push(
    `    start_path: ${JSON.stringify(w.startPath)}`,
    '    expect:',
    `      - ${JSON.stringify(w.expect)}`,
    '    budget:',
    `      steps: ${w.steps}`,
    '    tags:',
    '      - discovered',
  )
  return lines.join('\n')
}

/**
 * The description: what to do, in sentences, and then the route.
 *
 * The route is in it on purpose and it is accessible names rather than
 * selectors. A description saying "click #signup-btn" breaks when the markup
 * changes; an accessible name survives a redesign and disappears when somebody
 * removes the label a screen reader depends on, which is the failure a workflow
 * should have.
 */
function describe(goal: string, seed: string | null, reached: boolean, journey: Move[]): string {
  let out = sentence(goal)
  out += seed ? ` Found by an exploration from seed ${seed}` : ' Found by an exploration'
  if (journey.length === 0) return `${out}, which did not get anywhere.`
  out += reached ? ', which got there by: ' : ', which walked: '
  const shown = journey.slice(0, MAX_JOURNEY_LINES)
  out += shown.map((m) => lowerFirst(moveSentence(m))).join(', ')
  if (journey.length > shown.length) {
    out += `, and ${journey.length - shown.length} more steps`
  }
  return `${out}.`
}

function moveSentence(m: Move): string {
  switch (m.kind) {
    case 'goto':
      return `Open ${pathOf(m.url ?? '/')}`
    case 'fill':
      return `Fill ${m.field ?? 'a field'}`
    case 'click':
      return `Press ${JSON.stringify(m.control ?? 'a control')}`
    default:
      return m.kind
  }
}

/**
 * Where the exploration began, as a path.
 *
 * From the journey's first goto rather than from the goal, because the journey
 * carries the URL the browser was actually on and a goal whose start path
 * redirected somewhere else would compile a workflow that starts in the wrong
 * place. Every URL in a journey carries the preview environment's host, which
 * is different on every run and belongs to nobody's repository, so it is
 * reduced to a path: a manifest with 127.0.0.1:8731 in it reads as a mistake
 * within a day.
 */
function startPath(journey: Move[]): string {
  for (const m of journey) {
    if (m.kind === 'goto' && m.url) return pathOf(m.url)
  }
  return '/'
}

function pathOf(raw: string): string {
  try {
    // A relative base, so a path that is already relative parses and an
    // absolute one loses its host. A URL the parser refuses is not worth
    // guessing at, and "/" is where a workflow with no start path begins.
    const u = new URL(raw, 'http://environment.invalid')
    return u.pathname === '' ? '/' : u.pathname + u.search
  } catch {
    return '/'
  }
}

/** The six kinds an exploration reports, in the words a person reads. Unknown
 *  passes through, because the engine can add one by releasing and a control
 *  plane that refused it would drop a finding rather than name it. */
function friction(kind: string): string {
  switch (kind) {
    case 'no_effect':
      return 'a control that did nothing'
    case 'dead_end':
      return 'a dead end'
    case 'revisit':
      return 'a path that loops back'
    case 'unnamed_control':
      return 'an unnamed control'
    case 'slow_response':
      return 'something slow to answer'
    default:
      return kind
  }
}

function slugify(name: string): string | null {
  const s = name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 63)
    .replace(/-+$/g, '')
  return /^[a-z0-9]/.test(s) ? s : null
}

function sentence(s: string): string {
  const t = s.trim()
  if (t === '') return t
  return /[.!?]$/.test(t) ? t : `${t}.`
}

function lowerFirst(s: string): string {
  return s === '' ? s : s.slice(0, 1).toLowerCase() + s.slice(1)
}

// ---------------------------------------------------------------------------
// Reading a document this process did not write
// ---------------------------------------------------------------------------

function object(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {}
}

function text(value: unknown, max: number): string | null {
  if (typeof value !== 'string') return null
  const trimmed = value.trim()
  if (trimmed.length === 0 || trimmed.length > max) return null
  return trimmed
}

function strings(value: unknown, maxItems: number, maxLength: number): string[] {
  if (!Array.isArray(value)) return []
  const out: string[] = []
  for (const item of value.slice(0, maxItems)) {
    const s = text(item, maxLength)
    if (s !== null) out.push(s)
  }
  return out
}

/** The journey, skipping what cannot be read. One malformed move must not
 *  discard the walk: a description that lost every step is far worse than one
 *  that lost the third. */
function moves(value: unknown): Move[] {
  if (!Array.isArray(value)) return []
  const out: Move[] = []
  for (const item of value.slice(0, 400)) {
    const m = object(item)
    const kind = text(m.kind, 40)
    if (kind === null) continue
    out.push({
      kind,
      ...(text(m.url, 2048) ? { url: text(m.url, 2048)! } : {}),
      ...(text(m.field, 300) ? { field: text(m.field, 300)! } : {}),
      ...(text(m.control, 300) ? { control: text(m.control, 300)! } : {}),
    })
  }
  return out
}

function findingList(value: unknown): Finding[] {
  if (!Array.isArray(value)) return []
  const out: Finding[] = []
  for (const item of value.slice(0, 200)) {
    const f = object(item)
    const kind = text(f.kind, 60)
    if (kind === null) continue
    out.push({
      kind,
      ...(text(f.url, 2048) ? { url: text(f.url, 2048)! } : {}),
      ...(text(f.control, 300) ? { control: text(f.control, 300)! } : {}),
    })
  }
  return out
}
